File: src/Andromeda/AvaloniaApp/ViewModels/Widgets/DownloadWidgetViewModel.cs
using Andromeda.Core.FSharp;
using Avalonia.Controls;
using ReactiveUI;
using ReactiveUI.Legacy;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Reactive;
using System.Threading.Tasks;
using System.Timers;

using Andromeda.AvaloniaApp.Helpers;
using Andromeda.AvaloniaApp.ViewModels.Windows;

namespace Andromeda.AvaloniaApp.ViewModels.Widgets {
    public class DownloadWidgetViewModel : SubViewModelBase {
        private readonly Queue<InstallationInfos> downloadQueue = new Queue<InstallationInfos>();

        private readonly IReactiveList<DownloadStatus> downloads = new ReactiveList<DownloadStatus>();
        public IReactiveList<DownloadStatus> Downloads { get => this.downloads; }

        public DownloadWidgetViewModel(Control control, ViewModelBase parent) : base(control, parent) {
            // Search regularly for upgrades
            this.UpgradeAllGames();
            var timer = new Timer(1 * 3600 * 1000);
            timer.AutoReset = true;
            timer.Elapsed += (a, b) => this.UpgradeAllGames();
            timer.Start();
        }

        // Overwrite AppData Setter to ensure that InstalledGames are always up-to-date
        protected override void SetAppData(DomainTypes.AppData appData) {
            base.SetAppData(appData);
            if (((MainWindowViewModel)this.Parent).InstalledGames != null) {
                ((MainWindowViewModel)this.Parent).InstalledGames.Clear();
                ((MainWindowViewModel)this.Parent).InstalledGames.AddRange(appData.installedGames);
            }
        }

        public void UpgradeAllGames() {
            if (this.AppData.authentication.IsAuth) {
                this.SetAppData(Installed.searchInstalled(this.AppData));
                var result = Installed.checkAllForUpdates(this.AppData);
                this.SetAppData(result.Item2);

                var list = result.Item1.ToList();
                var message = "Found " + list.Count() + " games to update.";
                Logger.LogInfo(message);
                ((MainWindowViewModel)this.Parent).AddNotification(message);
                foreach (var updateInfo in list) {
                    var game = this.AppData.installedGames.Where(g => g.id == updateInfo.game.id).FirstOrDefault();
                    Debug.Assert(game != null);
                    if (updateInfo.newVersion != game.version) // Just to be sure
                    {
                        var result2 = Games.getAvailableInstallersForOs(this.AppData, game.id);
                        this.SetAppData(result2.Item2);

                        var installerInfo = result2.Item1.ToList().First();
                        AddDownload(new InstallationInfos(game.name, installerInfo));
                    }
                }
            }
        }

        public void AddDownload(InstallationInfos info) {
            this.downloadQueue.Enqueue(info);
            Logger.LogInfo("Added download of " + info.GameTitle + " to download queue.");

            this.CheckForNewDownload();
        }

        private void CheckForNewDownload() {
            // TODO: Check for nr of already running downloads
            if (this.downloadQueue.Count > 0) {
                this.StartDownload(this.downloadQueue.Dequeue());
            }
        }

        private void StartDownload(InstallationInfos info) {
            Logger.LogInfo("Get download info for " + info.GameTitle + " to download queue.");
            var res = Games.downloadGame(this.AppData, info.GameTitle, info.InstallerInfo);
            // TODO: This is installation stuff, this should be moved to core
            if (res != null) {
                var filepath = res.Value.Item2;
                var tmppath = res.Value.Item3;
                var downloadInfo = new DownloadStatus(info.GameTitle, tmppath, res.Value.Item4 / 1000000.0f);
                this.Downloads.Add(downloadInfo);

                Timer timer = null;
                Task downloadTask = null;

                // File not found in cache
                if (res.Value.Item1 != null) {
                    Logger.LogInfo("Download installer for " + info.GameTitle + ".");
                    downloadTask = res.Value.Item1.Value;
                    timer = new Timer(500.0);
                    timer.AutoReset = true;
                    timer.Elapsed += (a, b) =>
                    {
                        var fileInfo = new FileInfo(downloadInfo.FilePath);
                        downloadInfo.UpdateDownloaded(fileInfo.Length / 1000000.0f);
                    };
                    timer.Start();
                }
                else {
                    Logger.LogInfo("Use cached installer for " + info.GameTitle + ".");
                }

                var worker = new BackgroundWorker();
                worker.DoWork += (arg, arg2) =>
                {
                    if (downloadTask != null) {
                        downloadTask.Wait();
                        File.Move(tmppath, filepath);
                    }
                    downloadInfo.FilePath = filepath;

                    if (timer != null) {
                        timer.Stop();
                    }

                    // Install game
                    downloadInfo.IndicateInstalling();
                    if (downloadInfo.FilePath != null) {
                        Logger.LogInfo("Unpack " + downloadInfo.GameTitle + " from " + downloadInfo.FilePath);
                        Games.extractLibrary(this.AppData, downloadInfo.GameTitle, downloadInfo.FilePath);
                        Logger.LogInfo(downloadInfo.GameTitle + " unpacked successfully!");
                    }
                    else {
                        Logger.LogError("Filepath to installer is empty! Something went wrong...");
                    }
                };
                worker.RunWorkerCompleted += (arg, arg2) =>
                {
                    this.SetAppData(Installed.searchInstalled(this.AppData));
                    this.Downloads.Remove(downloadInfo);
                    Logger.LogInfo("Cleaned up after install.");
                };
                worker.RunWorkerAsync();
            }
        }
    }
}

File: src/Andromeda/AvaloniaApp.FSharp/ViewModels/DownloadWidgetViewModel.fs
namespace Andromeda.AvaloniaApp.FSharp.ViewModels

open Andromeda.Core.FSharp
open Andromeda.Core.FSharp.Installed
open GogApi.DotNet.FSharp.Base
open ReactiveUI.Legacy
open System
open System.Collections.Generic
open System.ComponentModel
open System.Diagnostics
open System.IO

open Andromeda.AvaloniaApp.FSharp.Helpers
open Andromeda.AvaloniaApp.FSharp.ViewModels

type DownloadWidgetViewModel(control, appDataWrapper) as this =
    inherit SubViewModelBase(control, appDataWrapper)

    let downloadQueue = Queue<InstallationInfos>()
    let downloads = ReactiveList<DownloadStatus>()

    member __.Init () =
        base.Init()
        this.UpgradeAllGames ()
        let timer = new Timers.Timer(1.0 * 3600.0 * 1000.0);
        timer.AutoReset <- true
        timer.Elapsed.Add(fun (_) -> this.UpgradeAllGames())
        timer.Start ()

    member val Downloads = downloads

    member __.UpgradeAllGames () =
        match this.AppData.authentication with
        | NoAuth -> ()
        | Auth _ ->
            this.AppData |> searchInstalled |> this.SetAppData
            let (list, appData) = this.AppData |> checkAllForUpdates
            appData |> this.SetAppData

            let message = "Found " + list.Length.ToString() + " games to update."
            message |> Logger.LogInfo // Move logging to Notification handling
            this.GetRootViewModel().AddNotification(message)

            for updateInfo in list do
                let games = this.AppData.installedGames |> List.where (fun game -> game.id = updateInfo.game.id)
                games.Length > 0 |> Debug.Assert
                // TODO: Select if multiple hits
                let game::_ = games
                match updateInfo.newVersion = game.version with
                | true -> ()
                | false ->
                    let (installerInfos, appData) = Games.getAvailableInstallersForOs appData game.id
                    appData |> this.SetAppData
                    // TODO: Select if multiple hits
                    let installerInfo::_ = installerInfos
                    InstallationInfos (game.name, installerInfo) |> this.AddDownload

    member this.AddDownload info =
        info |> downloadQueue.Enqueue
        "Added download of " + info.GameTitle + " to download queue." |> Logger.LogInfo

        this.CheckForNewDownload ()

    member this.CheckForNewDownload () =
        match downloadQueue.Count > 0 with
        | true -> downloadQueue.Dequeue() |> this.StartDownload
        | false -> ()

    member this.StartDownload info =
        "Get download info for " + info.GameTitle + " to download queue." |> Logger.LogInfo
        match Games.downloadGame this.AppData info.GameTitle info.InstallerInfo with
        | None -> ()
        | Some (task, filepath, tmppath, size) ->
            let downloadInfo = DownloadStatus(info.GameTitle, tmppath, float(size) / 1000000.0)
            this.Downloads.Add(downloadInfo)

            let taskData =
                match task with
                | Some task ->
                    "Download installer for " + info.GameTitle + "." |> Logger.LogInfo
                    let timer = new Timers.Timer (500.0)
                    timer.AutoReset <- true
                    timer.Elapsed.Add(fun _ ->
                        downloadInfo.FilePath
                        |> FileInfo
                        |> fun fileInfo -> int(float(fileInfo.Length) / 1000000.0)
                        |> downloadInfo.UpdateDownloaded
                    )
                    timer.Start()
                    Some (task, timer)
                | None ->
                    "Use cached installer for " + info.GameTitle + "." |> Logger.LogInfo
                    None

            let worker = new BackgroundWorker ()
            worker.DoWork.Add(fun _ ->
                match taskData with
                | Some (downloadTask, timer) ->
                    downloadTask.Wait()
                    File.Move(tmppath, filepath)
                    timer.Stop()
                | None -> ()

                downloadInfo.FilePath <- filepath

                // Install game
                downloadInfo.IndicateInstalling()
                "Unpack " + downloadInfo.GameTitle + " from " + downloadInfo.FilePath |> Logger.LogInfo
                Games.extractLibrary this.AppData downloadInfo.GameTitle downloadInfo.FilePath
                downloadInfo.GameTitle + " unpacked successfully!" |> Logger.LogInfo
            )
            worker.RunWorkerCompleted.Add(fun _ ->
                downloadInfo |> this.Downloads.Remove |> ignore
                searchInstalled this.AppData |> this.SetAppData
                "Cleaned up after install." |> Logger.LogInfo
            )
            worker.RunWorkerAsync ()
